modules/core/src/main/java/org/apache/ignite/internal/client/thin/ReliableChannel.java

```diff
@@ -487,10 +490,10 @@ private void rollCurrentChannel(ClientChannelHolder hld) {
      * On current channel failure.
      */
     private void onChannelFailure(ClientChannel ch, Throwable t, @Nullable List<ClientConnectionException> failures) {
-        // There is nothing wrong if curChIdx was concurrently changed, since channel was closed by another thread
+        // There is nothing wrong if defaultChIdx was concurrently changed, since channel was closed by another thread
         // when current index was changed and no other wrong channel will be closed by current thread because
         // onChannelFailure checks channel binded to the holder before closing it.
-        onChannelFailure(channels.get(curChIdx), ch, t, failures);
+        onChannelFailure(channels.get(lastDefaultChIdx), ch, t, failures);
```

Comment:
This won't work properly anymore, since we use a new index for every operation.

`channels.get(lastDefaultChIdx)` won't match `ch` in most cases, so `hld.closeChannel();` won't be called inside `onChannelFailure`.

I think we should rework failure scenario a bit:
* Get rid of `lastDefaultChIdx`, new `curChIdx` is enough
* Remove `ClientChannel ch` parameter from `onChannelFailure`. Use `nodeChannels` to resolve `ClientChannelHolder` from `ClientChannel`.